rust/src/dataset.rs

```diff
@@ -820,6 +820,44 @@ impl Dataset {
             Ok(vec![])
         }
     }
+
+    /// Delete the Dataset and all its data files.
+    ///
+    /// WARNING This operation cannot be undone.
+    pub async fn delete_dataset(&self) -> Result<()> {
+        // Delete  data / manifest files for all versions
+        for version in self.versions().await? {
+            self.delete_dataset_version(version).await?;
+        }
```

Comment:
Why do we delete specific files?

IMO I would write this like:

```rust
if self.object_store.is_local() {
   std::fs::remove_dir_all(self.base_dir)?;
} else {
   let locations = store.list(None).await?
        .map(|meta: Result<ObjectMeta, _>| meta.map(|m| m.location))
        .boxed();

    store.delete_stream(locations).try_collect::<Vec<Path>>().await?;
}
```

[delete_stream](https://docs.rs/object_store/0.6.1/object_store/trait.ObjectStore.html#method.delete_stream) is the most efficient way to delete many files in `object_store`. It will use bulk delete APIs when available. For example, in S3 we can delete 1,000 objects in one request. This will also handle making requests concurrently.

Reply:
Well as we iterate in the format we’ll be adding files so this method will need to keep up with that.